### ISSUE TITLE
Add optimal layout synthesis to Cirq

### ISSUE THREAD

**Issue by tanujkhattar** [quantumlib/Cirq#5855]
**Is your feature request related to a use case or problem? Please describe.**
Cirq now has [abstractions for qubit routing](https://github.com/quantumlib/Cirq/tree/master/cirq-core/cirq/transformers/routing) and an implementation of CQCs qubit routing algorithm as described in https://arxiv.org/abs/1902.08091

There is more recent work on using SMT solvers for qubit routing which has a significantly better performance in terms of number of swaps inserted, compared to CQCs heuristic routing algorithm. This work is described in https://arxiv.org/abs/2007.15671 and a corresponding implementation exists in https://github.com/tbcdebug/OLSQ

It would be nice to either add an implementation of the algorithm described in the paper to Cirq; or have a converter to/from Cirq to the OLSQ solver so that cirq users can easily route their cirq circuits using this algorithm.

This would be a good independent project for anyone who is interested in learning more about qubit routing. 

**What is the urgency from your perspective for this issue? Is it blocking important work?**
P3 - I'm not really blocked by it, it is an idea I'd like to discuss / suggestion based on principle

**Comment by anonymousr007** [comment]
Hello @tanujkhattar, I want to work on this issue as a part-time project. Can you please assign me this issue? We discuss this issue very briefly. Thanks.

**Comment by senecameeks** [comment]
@tanujkhattar can you assign to me? I think this would be a great learning opportunity :)

**Comment by dstrain115** [comment]
Cirq cync:  accepted but low priority.  Seneca will pick another issue, so this is open for volunteers.

**Comment by Prashik123** [comment]
Hi @tanujkhattar and @mhucka! I noticed this issue recently received the `help wanted` tag, and I would love to take it up as a part-time project. 

I've been looking into qubit routing optimization and have read through the OLSQ paper (arXiv:2007.15671). Given the two options proposed in the issue description, I think the most effective first step would be to build a robust converter interface between `cirq.Circuit` and the existing `OLSQ` solver. 

This approach would immediately allow Cirq users to leverage the significantly better swap performance of SMT-based routing, while keeping the initial PR scoped and manageable. Once the converter is merged and stable, we could always revisit whether a native implementation inside Cirq is worth the extra overhead.

Does this direction sound good to you? If so, please feel free to assign this to me, and I will get started on a draft PR!

**Comment by mhucka** [comment]
@Prashik123 Thank you for your interest!

As far the approach, a disadvantage of targeting an external third-party software library like [tbcdebug/OLSQ](https://github.com/tbcdebug/OLSQ) is that it would require Cirq to add it as a dependency. Adding another dependency is not _necessarily_ always a problem, but the implications need to be considered carefully. The external package has to be something that is robust and actively maintained, or else it risks becoming a burden to the Cirq maintainers. In the case of tbcdebug/OLSQ, the repository does not seem to have been active for 4 years, which does not inspire confidence about its future.

So while I understand the desire to avoid reinventing wheels, in this case, it looks safer to assume that a native Cirq implementation would be the preferred end goal and that we would want to avoid relying on the OLSQ implementation. Whether creating a converter from Cirq to OLSQ makes sense as a stepping stone, I can't say – maybe it would help learn how everything works before diving into the (probably harder) task of implementing the algorithm natively in Cirq, or maybe it would turn out to be a waste of time.

In any case, I'll go ahead and assign the issue to you. Thanks again.
